glean-core/benches/dispatcher.rs

```diff
@@ -0,0 +1,111 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Micro-benchmark to measure the time it takes to launch tasks on the dispatcher.
+//! Explicitly does not measure the time it takes to _run_ these.
+
+use std::hint::black_box;
+
+use criterion::{criterion_group, criterion_main, Criterion};
+use glean_core::{
+    dispatcher, glean_initialize, glean_shutdown, join_init, ClientInfoMetrics, CommonMetricData,
+    CounterMetric, Lifetime, OnGleanEvents,
+};
+
+pub fn dispatcher_benchmark(c: &mut Criterion) {
+    // Ensure dispatcher has been created and is usable.
+    dispatcher::flush_init().unwrap();
+
+    // An empty function has size 0. Putting that on the queue should be quick.
+    c.bench_function("empty fn", |b| {
+        b.iter(|| {
+            dispatcher::launch(|| {
+                // intentionally left empty
+            });
+        })
+    });
+
+    // Drain the dispatcher.
+    dispatcher::block_on_queue();
+
+    // Moving data into the function means we need to copy that data on to the queue too.
+    c.bench_function("fn 1024", |b| {
+        b.iter(|| {
+            let c = [0; 1024];
+            dispatcher::launch(move || {
+                black_box(c);
+            });
+        })
+    });
+
+    // Ensure we can run more tests afterwards.
+    dispatcher::reset_dispatcher();
+}
+
+struct Callbacks;
+
+impl OnGleanEvents for Callbacks {
+    fn initialize_finished(&self) {}
+
+    fn trigger_upload(&self) -> glean_core::Result<(), glean_core::CallbackError> {
+        Ok(())
+    }
+
+    fn start_metrics_ping_scheduler(&self) -> bool {
+        false
+    }
+
+    fn cancel_uploads(&self) -> glean_core::Result<(), glean_core::CallbackError> {
+        Ok(())
+    }
+}
+
+pub fn metric_dispatcher_benchmark(c: &mut Criterion) {
+    let dir = tempfile::tempdir().unwrap();
+    let data_path = dir.path().display().to_string();
+
+    let cfg = glean_core::InternalConfiguration {
```

Comment:
Oh no, do we now need an _internal_ configurationbuilder?

Reply:
Well, technically this is also not public API, so other than our wrappers no one should use this. Just for the benchmarking purposes we reach to the lowest level.

Other benchmarks might as well then run on the RLB and use the public APIs.